File: Flutter/lib/models/poll.dart
class Poll {
  final int pollId;
  final String question;
  final List<PollOption> options;
  final int totalVotes;
  final DateTime? endsAt;
  final bool hasVoted;
  final List<int>? userVoteOptionIndexes;
  final bool isMultipleChoice;
  final bool allowChangeVote;
  final bool showResultsBeforeVote;
  final String? imageUrl;

  Poll({
    required this.pollId,
    required this.question,
    required this.options,
    this.totalVotes = 0,
    this.endsAt,
    this.hasVoted = false,
    this.userVoteOptionIndexes,
    this.isMultipleChoice = false,
    this.allowChangeVote = true,
    this.showResultsBeforeVote = true,
    this.imageUrl,
  });

  factory Poll.fromJson(Map<String, dynamic> json) {
    return Poll(
      pollId: json['pollId'] ?? 0,
      question: json['question'] ?? '',
      options:
          (json['options'] as List<dynamic>?)
              ?.map((o) => PollOption.fromJson(o))
              .toList() ??
          [],
      totalVotes: json['totalVotes'] ?? 0,
      endsAt: json['endsAt'] != null ? DateTime.parse(json['endsAt']) : null,
      hasVoted: (json['userSelections'] != null) && ((json['userSelections'] as List).isNotEmpty),
      userVoteOptionIndexes: (json['userSelections'] as List<dynamic>?)?.map((e) => (e as num).toInt()).toList(),
      isMultipleChoice: json['isMultipleChoice'] ?? false,
      allowChangeVote: json['allowChangeVote'] ?? true,
      showResultsBeforeVote: json['showResultsBeforeVote'] ?? true,
      imageUrl: json['imageUrl'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'pollId': pollId,
      'question': question,
      'options': options.map((o) => o.toJson()).toList(),
      'totalVotes': totalVotes,
      'endsAt': endsAt?.toIso8601String(),
      'hasVoted': hasVoted,
      'userSelections': userVoteOptionIndexes,
      'isMultipleChoice': isMultipleChoice,
      'allowChangeVote': allowChangeVote,
      'showResultsBeforeVote': showResultsBeforeVote,
      'imageUrl': imageUrl,
    };
  }

  bool get isExpired {
    if (endsAt == null) return false;
    return DateTime.now().isAfter(endsAt!);
  }
}

class PollOption {
  final String text;
  final int voteCount;
  final double percentage;

  PollOption({required this.text, this.voteCount = 0, this.percentage = 0.0});

  factory PollOption.fromJson(Map<String, dynamic> json) {
    return PollOption(
      text: json['text'] ?? json['optionText'] ?? '',
      voteCount: json['voteCount'] ?? 0,
      percentage: (json['percentage'] ?? 0.0).toDouble(),
    );
  }

  Map<String, dynamic> toJson() {
    return {'text': text, 'voteCount': voteCount, 'percentage': percentage};
  }
}

File: API/Models/Poll.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace InstapropAPI.Models
{
    public class Poll
    {
        [Key]
        public long PollId { get; set; }

        [Required]
        public long PostId { get; set; }

        [ForeignKey(nameof(PostId))]
        public CommunityPost Post { get; set; } = null!;

        [Required]
        public string Question { get; set; } = string.Empty;

        // JSON array of {optionText, voteCount}
        [Column(TypeName = "TEXT")]
        [Required]
        public string Options { get; set; } = string.Empty;

        public DateTime? EndsAt { get; set; }

        public int TotalVotes { get; set; } = 0;

        // Flags
        public bool IsMultipleChoice { get; set; } = false;
        public bool AllowChangeVote { get; set; } = true;
        public bool ShowResultsBeforeVote { get; set; } = true;

        // Optional image for the poll question
        [MaxLength(500)]
        public string? ImageUrl { get; set; }

        // Relations
        public ICollection<PollVote> Votes { get; set; } = new List<PollVote>();
    }

    public class PollVote
    {
        [Key]
        public long VoteId { get; set; }

        [Required]
        public long PollId { get; set; }

        [ForeignKey(nameof(PollId))]
        public Poll Poll { get; set; } = null!;

        [Required]
        public long AccountId { get; set; }

        [ForeignKey(nameof(AccountId))]
        public Account Account { get; set; } = null!;

        [Required]
        public int OptionIndex { get; set; } // Which option was selected

        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    }
}

